smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/RequestCompressionTraitValidator.java

```diff
@@ -0,0 +1,102 @@
+ * - Validate encodings has at least one compression algorithm
+ * - Validate encodings are all supported compression algorithms
+ */
+public final class RequestCompressionTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        Set<OperationShape> shapes = model.getOperationShapesWithTrait(RequestCompressionTrait.class);
+        shapes.forEach(shape -> {
+            validateOperationInput(model, shape, events);
+            validateEncodings(model, shape, events);
+        });
+        return events;
+    }
+
+    private void validateOperationInput(Model model, OperationShape operationShape, List<ValidationEvent> events) {
+        // Validate operation input has no member with both streaming and requiresLength
+        // traits applied
+        if (operationShape.getInputShape().equals(ShapeId.from("smithy.api#Unit"))) {
+            events.add(error(operationShape,
+                    "The requestCompression trait can only be applied to operations which have an "
+                            + "input shape defined"));
+            return;
+        }
+        StructureShape inputShape = model.expectShape(operationShape.getInputShape(), StructureShape.class);
+        for (Map.Entry<String, MemberShape> entry : inputShape.getAllMembers().entrySet()) {
+            MemberShape memberShape = entry.getValue();
+            Shape targetShape = model.expectShape(memberShape.getTarget());
+            if (targetShape.hasTrait(StreamingTrait.class)
+                    && targetShape.hasTrait(RequiresLengthTrait.class)) {
+                events.add(error(operationShape, String.format(
+                        "The requestCompression trait can only be applied to operations which do not have "
+                                + "input members that target shapes with both the streaming and requiresLength"
+                                + "trait applied, but found member `%s` targeting `%s`",
+                        memberShape.getId(), targetShape.getId())));
+            }
+        }
+    }
+
+    private void validateEncodings(Model model, OperationShape operationShape, List<ValidationEvent> events) {
+        // Validate encodings have at least one compression algorithm
+        RequestCompressionTrait trait = operationShape.expectTrait(RequestCompressionTrait.class);
+        if (trait.getEncodings().isEmpty()) {
+            events.add(error(operationShape, trait, String.format(
+                    "there must be at least one compression algorithm in `encodings` for the "
+                            + "requestCompression trait applied to `%s`",
+                    operationShape.getId())));
+            return;
+        }
+        // Validate encodings are all supported compression algorithms
+        List<String> invalidEncodings = new ArrayList<>();
+        for (String encoding : trait.getEncodings()) {
```

Comment:
I think if there was a "supportsEncoding" method (see below) then this would look better. 
Then it would be: 
```
List<String> invalidEncodings = traits.getEncodings.stream()
    .filter(RequestCompressionTrait::supportsEncoding)
    .collect(Collectors.toList());
```
Which I like, but just a matter of taste 🤷